crates/wysiwyg/src/dom/insert_node_at_range.rs

```diff
@@ -0,0 +1,172 @@
+// Copyright 2023 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::{DomHandle, DomNode, UnicodeString};
+
+use super::{Dom, Range};
+
+impl<S> Dom<S>
+where
+    S: UnicodeString,
+{
+    pub fn insert_node_at_range(
+        &mut self,
+        range: &Range,
+        new_node: DomNode<S>,
+    ) -> DomHandle {
+        if range.is_cursor() {
+            self.insert_node_at_cursor(range, new_node)
+        } else {
+            self.insert_node_at_selection(range, new_node)
+        }
+    }
+
+    fn insert_node_at_cursor(
+        &mut self,
+        range: &Range,
+        new_node: DomNode<S>,
+    ) -> DomHandle {
```

Comment:
Currently this method could be called with a range that is not a cursor. Is there any way we can make this function safer and not allow that to happen? I'm thinking you could change it to accept the cursor location for example:

```rust
    fn insert_node_at_cursor(
        &mut self,
        cursor: Option<&Location>,
        new_node: DomNode<S>,
    )
```

Reply:
I don't think it can as I think the only exposed function is `insert_node_at_range` which then checks if it's a cursor before calling `insert_node_at_cursor`. Could be wrong though.